# TWC Case Study Analysis

## Table of Contents
<ul>
<li><a href="#intro">Introduction</a></li>
<li><a href="#wrangling">Data Wrangling</a></li>
<li><a href="#eda">Exploratory Data Analysis</a></li>
<li><a href="#conclusions">Conclusions</a></li>
</ul>

<a id='intro'></a>
## Introduction
TWC case study is a sample medical dataset of patient interaction.

Using the TWC Case study dataset, analysis will be carried out with the aim of finding the trends. 

To achieve this, the following questions will be answered in the analysis:

- Which Patient Type in Location 10 brought the Most revenue in 2019? 
- Which location saw the biggest increase in total patient volume %? what %?  (1/1/2020-3/16/2020  vs 3/17/2020-6/30/2020).
- Do females or males bring in more revenue/patient?
- What is the cost/patient of our most costly patient type? Which patient type?
- Which Physician was the most profitable on a per patient basis? What profit?
- Which doctor saw the highest volume of unique Medicare Patients.
- What is the average age of a female Glaucoma & Glaucoma Laser patient in 2019?
- Which patient type drives the most revenue/encounter and profit/encounter? What was Revenue by Encounter/ Profit/Encounter?
- Which contract type is most profitable per patient/what profit?
- What is the total profit on FFS patients?


To answer the above questions from the TWC case study dataset Dataset will be used.



In [303]:
#Import needed libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import calendar

%matplotlib inline


<a id='wrangling'></a>
## Data Wrangling

TWC Case Study dataset loading and cleaning operations

### General Properties

In [2]:
# Load TWC case study dataset
df_data = pd.read_excel("TWC_Case Study dataset.xlsx", sheet_name="Data")
df_cpt_Info = pd.read_excel("TWC_Case Study dataset.xlsx", sheet_name="CPT Info")
df_contract_info = pd.read_excel("TWC_Case Study dataset.xlsx", sheet_name="Contract Info")
df_cost_mca_rev = pd.read_excel("TWC_Case Study dataset.xlsx", sheet_name="Cost as % of MCA or Rev")

In [3]:
# Confirm Import 
df_data.head(2)

Person Number charge DOS    Contract    COB1    Location    Physician  \
0          21913 2020-01-06  Contract 1  COB1 4  Location 4  Physician 4   
1          11093 2020-01-02  Contract 1  COB1 6  Location 6  Physician 7   

   Patient Age  Cpt Type  Co-Payment  Encounter Number        Patient Type  \
0           73  Lucentis           0                 5  Retinal Injections   
1           80     Eylea           0                15  Retinal Injections   

   Patient Zip Code Patient Gender         CPT  Revenue  
0           90011.0              M  LUCEN_DELV  1823.64  
1           90044.0              M  EYLEA_DELV  2447.55

In [4]:
df_cpt_Info.head(2)

Code Desc  Cpt Type Service Item Id  Medicare Allowable
0  Lucentis Delive  Lucentis      LUCEN_DELV                 0.0
1  Eylea Delivered     Eylea      EYLEA_DELV                 0.0

In [5]:
df_contract_info.head(2)

Number    Contract  Number.1   COB 1 Fin Class       LOB  Number.2  TPA  \
0       1  Contract 1         4  COB1 4       FFS  Medicare       NaN  NaN   
1       1  Contract 1         6  COB1 6       FFS  Medicare       NaN  NaN   

        Primary Key  Countif  
0  Contract 1COB1 4        1  
1  Contract 1COB1 6        1

In [6]:
df_cost_mca_rev.head(2)

Cpt Type  % of MCA  % of Revenue
0  Lucentis      0.65           0.9
1     Eylea      0.65           0.9

## Assessing Data
Assessing collected dataset

##### Visual assessment:

In [7]:
df_data.sample(10)

Person Number charge DOS     Contract      COB1     Location  \
1537          23046 2019-06-06  Contract 29  COB1 138  Location 23   
7330          21913 2019-03-07   Contract 1    COB1 4   Location 4   
6617          15805 2019-04-09  Contract 18   COB1 28  Location 31   
8353          10440 2019-02-14   Contract 1  COB1 106   Location 6   
1497          10922 2020-10-08  Contract 50   COB1 85   Location 4   
3232          12885 2018-05-04  Contract 21   COB1 30   Location 4   
3460          17201 2018-08-30   Contract 1  COB1 106   Location 6   
4179          26360 2020-08-17  Contract 63  COB1 113  Location 23   
1938          11322 2018-07-25  Contract 46   COB1 72  Location 14   
8890          16046 2020-01-16  Contract 16  COB1 108   Location 6   

         Physician  Patient Age          Cpt Type  Co-Payment  \
1537  Physician 14           75  Cataract Surgery           0   
7330   Physician 4           72               OCT           0   
6617  Physician 18           52       Office Code           0   
8353   Physician 7           75       Office Code           0   
1497   Physician 4           67               OCT           5   
3232   Physician 4           49             Other           0   
3460   Physician 7           86             Other           0   
4179  Physician 48           87        Anesthesia           0   
1938  Physician 16           60       Office Code           0   
8890   Physician 7           71               OCT           0   

      Encounter Number        Patient Type  Patient Zip Code Patient Gender  \
1537              7028             Surgery           90049.0              M   
7330              3573  Retinal Injections           90011.0              M   
6617             21237             Post Op           90065.0              M   
8353             21208  Retinal Injections           90017.0              M   
1497              7604             Post Op           90077.0              M   
3232              2089  Retinal Injections           90023.0              M   
3460             15175  Retinal Procedures           90036.0              F   
4179             18533          Anesthesia           90049.0              M   
1938              9487  Retinal Injections           90003.0              M   
8890             28434  Retinal Injections           90040.0              M   

        CPT     Revenue  
1537  66984  976.892000  
7330  92134   61.124000  
6617  2022F    0.000000  
8353  2022F    0.000000  
1497  92134   60.430447  
3232  67028  156.087570  
3460  67028  156.087570  
4179    140  183.859200  
1938  92014  182.383477  
8890  92134   60.430447

In [8]:
df_cpt_Info.sample(10)

Code Desc          Cpt Type Service Item Id  Medicare Allowable
74   Anes- Eye; Corn        Anesthesia             144          219.690000
3    Medical Records             Other            Misc            0.000000
66   Lucentis 0.5 Mg          Lucentis           J2778            0.000000
59   Cataract Ext.    Cataract Surgery           66982          815.531849
43   Fistuliz Scler;          Glaucoma           66170         1190.778929
106  Repr Blepharopt   Plastic Surgery           67904          594.498283
2    Walk Out Not Se             Other        Walk Out            0.000000
18   Ophth Serv: Med       Office Code           92012           90.520274
64   Ophth Exten W/r    Ophthalmoscopy           92226           26.390947
46   Inj Anes Agent;             Other           64400           88.401338

In [9]:
df_contract_info.sample(10)

Number      Contract  Number.1     COB 1 Fin Class         LOB  Number.2  \
38      66   Contract 66       119  COB1 119       CAP    Medicaid      16.0   
25       1    Contract 1        81   COB1 81       FFS    Medicare       NaN   
45      40   Contract 40       133  COB1 133       FFS    Medicare       NaN   
17      42   Contract 42        53   COB1 53       FFS    Medicare      11.0   
61      12   Contract 12        37   COB1 37       CAP    Medicare       6.0   
69     103  Contract 103       194  COB1 194       CAP    Medicaid       NaN   
71     109  Contract 109       201  COB1 201       CAP    Medicare       6.0   
2        5    Contract 5         9    COB1 9  Self-Pay  Commercial       NaN   
24      49   Contract 49        80   COB1 80       CAP    Medicare      13.0   
4       12   Contract 12        18   COB1 18       FFS    Medicare       6.0   

       TPA           Primary Key  Countif  
38  TPA 16   Contract 66COB1 119        1  
25     NaN     Contract 1COB1 81        1  
45     NaN   Contract 40COB1 133        1  
17  TPA 11    Contract 42COB1 53        2  
61   TPA 6    Contract 12COB1 37        2  
69     NaN  Contract 103COB1 194        1  
71   TPA 6  Contract 109COB1 201        1  
2      NaN      Contract 5COB1 9        1  
24  TPA 13    Contract 49COB1 80        1  
4    TPA 6    Contract 12COB1 18        1

In [10]:
df_cost_mca_rev.sample(10)

Cpt Type  % of MCA  % of Revenue
10       Plastic Surgery      0.95           0.9
8              SLT Laser      0.55           0.9
9           Visual Field      0.95           0.9
3                Avastin      0.65           0.9
16  Non-Cataract Surgery      0.80           0.9
18       Retinal Surgery      0.85           0.9
14              Glaucoma      0.75           0.9
4                    OCT      0.65           0.9
19            Anesthesia      0.50           0.9
0               Lucentis      0.65           0.9

##### Programmatic assessment:

In [11]:
# check the data table for data Quality issues
df_data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10320 entries, 0 to 10319
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Person Number     10320 non-null  int64         
 1   charge DOS        10320 non-null  datetime64[ns]
 2   Contract          10320 non-null  object        
 3   COB1              10320 non-null  object        
 4   Location          10320 non-null  object        
 5   Physician         10320 non-null  object        
 6   Patient Age       10320 non-null  int64         
 7   Cpt Type          10320 non-null  object        
 8   Co-Payment        10320 non-null  int64         
 9   Encounter Number  10320 non-null  int64         
 10  Patient Type      10320 non-null  object        
 11  Patient Zip Code  10121 non-null  float64       
 12  Patient Gender    10320 non-null  object        
 13  CPT               10320 non-null  object        
 14  Revenue           1032

In [12]:
df_cpt_Info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 130 entries, 0 to 129
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   Code Desc           130 non-null    object 
 1   Cpt Type            130 non-null    object 
 2   Service Item Id     130 non-null    object 
 3   Medicare Allowable  130 non-null    float64
dtypes: float64(1), object(3)
memory usage: 4.2+ KB


In [13]:
df_contract_info.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 83 entries, 0 to 82
Data columns (total 10 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   Number       83 non-null     int64  
 1   Contract     83 non-null     object 
 2   Number.1     83 non-null     int64  
 3   COB 1        83 non-null     object 
 4   Fin Class    83 non-null     object 
 5   LOB          83 non-null     object 
 6   Number.2     36 non-null     float64
 7   TPA          36 non-null     object 
 8   Primary Key  83 non-null     object 
 9   Countif      83 non-null     int64  
dtypes: float64(1), int64(3), object(6)
memory usage: 6.6+ KB


In [99]:
df_cost_mca_rev.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 23 entries, 0 to 22
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Cpt Type      23 non-null     object 
 1   % of MCA      23 non-null     float64
 2   % of Revenue  23 non-null     float64
dtypes: float64(2), object(1)
memory usage: 680.0+ bytes


In [101]:
df_cpt_Info[df_cpt_Info.duplicated('Service Item Id')]

Code Desc          Cpt Type Service Item Id  Medicare Allowable
94   Cataract Ext. C  Cataract Surgery           66982          815.531849
122  Beovu Delivered             Beovu      Beovu_Delv            0.000000

In [103]:
df_cost_mca_rev[df_cost_mca_rev.duplicated('Cpt Type')]

Empty DataFrame
Columns: [Cpt Type, % of MCA, % of Revenue]
Index: []

#### Quality issues
##### `df_data` table 
1. Erroneous datatypes (Person Number and Patient Zip Code should be objects, not integers or floats as they are not numeric and not intended for calculations)

2. Duplicates on Service Item Id in `df_cpt_Info`, COB 1 in `df_contract_info`

##### `df_cost_mca_rev` table
3. % of MCA and % of Revenue values not in percentage a

#### Tidiness issues
1. Based on tidiness rule that Each type of observational unit forms a table, Cost column should be generated from `df_cost_mca_rev` into `df_data`. Also Fin Class, LOB, and TPA from `df_contract_info` should be merged into `df_data`

### Data Cleaning

In [602]:
# Make copies of original pieces of data
df_data_clean = df_data.copy()
df_cpt_Info_clean = df_cpt_Info.copy()
df_contract_info_clean = df_contract_info.copy()
df_cost_mca_rev_clean = df_cost_mca_rev.copy()

#### Issue:
Erroneous datatypes

##### Define
- convert Person Number and Patient Zip to object format

##### Code

In [603]:
# change datatype
df_data_clean['Person Number'] = df_data_clean['Person Number'].astype(str)
df_data_clean['Patient Zip Code'] = df_data_clean['Patient Zip Code'].astype(str)

##### Test

In [604]:
#confirm type
df_data_clean.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10320 entries, 0 to 10319
Data columns (total 15 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Person Number     10320 non-null  object        
 1   charge DOS        10320 non-null  datetime64[ns]
 2   Contract          10320 non-null  object        
 3   COB1              10320 non-null  object        
 4   Location          10320 non-null  object        
 5   Physician         10320 non-null  object        
 6   Patient Age       10320 non-null  int64         
 7   Cpt Type          10320 non-null  object        
 8   Co-Payment        10320 non-null  int64         
 9   Encounter Number  10320 non-null  int64         
 10  Patient Type      10320 non-null  object        
 11  Patient Zip Code  10320 non-null  object        
 12  Patient Gender    10320 non-null  object        
 13  CPT               10320 non-null  object        
 14  Revenue           1032

#### Issue:
% of MCA and % of Revenue values not in percentage

##### Define
- Multiply % of MCA and % of Revenue values by 100 to convert to percentage

##### Code

In [605]:
df_cost_mca_rev_clean['% of MCA'] = df_cost_mca_rev_clean['% of MCA'] * 100
df_cost_mca_rev_clean['% of Revenue'] = df_cost_mca_rev_clean['% of Revenue'] * 100

##### Test

In [606]:
#confirm value in percentage 
df_cost_mca_rev_clean[['% of MCA','% of Revenue' ]].sample(5)

% of MCA  % of Revenue
0       65.0          90.0
8       55.0          90.0
1       65.0          90.0
2       65.0          90.0
11      65.0          90.0

#### Issue:
Duplicates on Service Item Id in df_cpt_Info

##### Define
- Drop duplicate detected and keep just one record

##### Code

In [607]:
df_cpt_Info_clean = df_cpt_Info_clean.drop_duplicates(subset=['Service Item Id'], keep = 'last')

##### Test

In [608]:
df_cpt_Info_clean[df_cpt_Info_clean.duplicated('Service Item Id')]

Empty DataFrame
Columns: [Code Desc, Cpt Type, Service Item Id, Medicare Allowable]
Index: []

#### Issue:
Duplicates on COB 1 in df_contract_info_clean

##### Define
- Drop duplicate detected and keep just one record

##### Code

In [609]:
df_contract_info_clean = df_contract_info_clean.drop_duplicates(subset=['COB 1'], keep = 'last')

##### Test

In [610]:
df_contract_info_clean[df_contract_info_clean.duplicated('COB 1')]

Empty DataFrame
Columns: [Number, Contract, Number.1, COB 1, Fin Class, LOB, Number.2, TPA, Primary Key, Countif]
Index: []

#### Issue:
Cost column should be generated from `df_cost_mca_rev` into `df_data`. Also Fin Class, LOB, and TPA from `df_contract_info` should be merged into `df_data`

##### Define
- represent Medicare Allowable value for each record in df_data
- generate the cost column in df_data using Revenue and Medicare Allowable based on the cost rule.

##### Code

In [611]:
#create the merging key on both tables
df_cpt_Info_clean['Key'] = df_cpt_Info_clean['Service Item Id']
df_data_clean['Key'] = df_data_clean['CPT']

In [612]:
#confirm match on first row
assert df_cpt_Info_clean['Key'][0] == df_data_clean['Key'][0]

In [613]:
# drop Cpt Type on df_cpt_Info_clean to avoid unique column error
df_cpt_Info_clean = df_cpt_Info_clean.drop(['Cpt Type'], axis = 1)

In [614]:
# merge the table with match on key column
df_data_clean = df_data_clean.merge(df_cpt_Info_clean, how='inner', on='Key')

In [615]:
# drop redudant columns
df_data_clean = df_data_clean.drop(['Key','Service Item Id'], axis = 1)

##### Test

In [616]:
#check new columns in df_data_clean
df_data_clean.columns

Index(['Person Number', 'charge DOS', 'Contract', 'COB1', 'Location',
       'Physician', 'Patient Age', 'Cpt Type', 'Co-Payment',
       'Encounter Number', 'Patient Type', 'Patient Zip Code',
       'Patient Gender', 'CPT', 'Revenue', 'Code Desc', 'Medicare Allowable'],
      dtype='object')

In [617]:
df_data_clean.sample(3)

Person Number charge DOS     Contract     COB1    Location     Physician  \
3826           817 2018-12-18  Contract 46  COB1 72  Location 4  Physician 14   
2217         12885 2019-03-21  Contract 21  COB1 30  Location 4   Physician 4   
221          13474 2018-12-27  Contract 42  COB1 53  Location 4  Physician 29   

      Patient Age     Cpt Type  Co-Payment  Encounter Number  \
3826           90  Office Code           0             15871   
2217           50  Office Code           0             21246   
221            89        Eylea           0              6765   

            Patient Type Patient Zip Code Patient Gender         CPT  \
3826        Office Visit          90041.0              M       92012   
2217  Retinal Injections          90023.0              M       2022F   
221   Retinal Injections          90006.0              F  EYLEA_DELV   

          Revenue        Code Desc  Medicare Allowable  
3826   129.905565  Ophth Serv: Med           90.520274  
2217     0.000000  DilatedRetinl E            0.000000  
221   2447.550000  Eylea Delivered            0.000000

In [618]:
# confirm that the number of unique values in CPT after the merge is still same
assert df_data_clean['CPT'].nunique() == df_data['CPT'].nunique()

In [619]:
# confirm the number of rows is same after merger
assert df_data.shape[0] == df_data_clean.shape[0]

#### Issue:
 Fin Class, LOB, and TPA from `df_contract_info` should be merged into `df_data`

##### Define
- merge contract info into data table.

##### Code

In [620]:
df_contract_info_clean.columns

Index(['Number', 'Contract', 'Number.1', 'COB 1', 'Fin Class', 'LOB',
       'Number.2', 'TPA', 'Primary Key', 'Countif'],
      dtype='object')

In [621]:
df_data_clean.columns

Index(['Person Number', 'charge DOS', 'Contract', 'COB1', 'Location',
       'Physician', 'Patient Age', 'Cpt Type', 'Co-Payment',
       'Encounter Number', 'Patient Type', 'Patient Zip Code',
       'Patient Gender', 'CPT', 'Revenue', 'Code Desc', 'Medicare Allowable'],
      dtype='object')

In [622]:
#create the merging key on both tables
df_contract_info_clean['Key'] = df_contract_info_clean['COB 1']
df_data_clean['Key'] = df_data_clean['COB1']

In [623]:
# drop Cpt Type on df_cpt_Info_clean to avoid unique column error
df_contract_info_clean = df_contract_info_clean.drop(['Contract'], axis = 1)

In [624]:
# merge the table with match on key column
df_data_clean = df_data_clean.merge(df_contract_info_clean, how='left', on='Key')

In [625]:
df_data_clean.columns

Index(['Person Number', 'charge DOS', 'Contract', 'COB1', 'Location',
       'Physician', 'Patient Age', 'Cpt Type', 'Co-Payment',
       'Encounter Number', 'Patient Type', 'Patient Zip Code',
       'Patient Gender', 'CPT', 'Revenue', 'Code Desc', 'Medicare Allowable',
       'Key', 'Number', 'Number.1', 'COB 1', 'Fin Class', 'LOB', 'Number.2',
       'TPA', 'Primary Key', 'Countif'],
      dtype='object')

In [626]:
# drop redudant columns
df_data_clean = df_data_clean.drop(['Number.2','Countif','Number', 'Number.1' ], axis = 1)

##### Test

In [627]:
#check new columns in df_data_clean
df_data_clean.columns

Index(['Person Number', 'charge DOS', 'Contract', 'COB1', 'Location',
       'Physician', 'Patient Age', 'Cpt Type', 'Co-Payment',
       'Encounter Number', 'Patient Type', 'Patient Zip Code',
       'Patient Gender', 'CPT', 'Revenue', 'Code Desc', 'Medicare Allowable',
       'Key', 'COB 1', 'Fin Class', 'LOB', 'TPA', 'Primary Key'],
      dtype='object')

In [628]:
df_data_clean.sample(3)

Person Number charge DOS    Contract      COB1     Location  \
2796         25436 2019-01-30  Contract 8   COB1 13  Location 14   
3159         17373 2019-05-03  Contract 8  COB1 147   Location 6   
6284         11093 2021-02-09  Contract 1    COB1 6   Location 6   

         Physician  Patient Age Cpt Type  Co-Payment  Encounter Number  ...  \
2796  Physician 25           57  Post-op           0             13023  ...   
3159  Physician 14           63  Post-op           0             23620  ...   
6284  Physician 16           81      OCT           0             28398  ...   

        CPT    Revenue        Code Desc Medicare Allowable       Key  \
2796  99024   0.000000  Postop F/u Visi           0.000000   COB1 13   
3159  99024   0.000000  Postop F/u Visi           0.000000  COB1 147   
6284  92134  55.915876  Scan Computeriz          43.084681    COB1 6   

         COB 1  Fin Class         LOB    TPA         Primary Key  
2796   COB1 13        CAP  Commercial  TPA 2   Contract 8COB1 13  
3159  COB1 147        CAP  Commercial  TPA 2  Contract 8COB1 147  
6284    COB1 6        FFS    Medicare    NaN    Contract 1COB1 6  

[3 rows x 23 columns]

In [629]:
# confirm that the number of unique values in CPT after the merge is still same
assert df_data_clean['COB1'].nunique() == df_data['COB1'].nunique()

In [630]:
# confirm the number of rows is same after merger
assert df_data.shape[0] == df_data_clean.shape[0]

#### Issue:
generate the cost column

##### Define
- generate the cost column in df_data using Revenue and Medicare Allowable based on the cost rule.

##### Code

In [631]:
# merge cost % records to df_data_clean table
df_data_clean = df_data_clean.merge(df_cost_mca_rev_clean, how='left', on='Cpt Type')

In [632]:
assert df_data_clean['Cpt Type'].nunique() == df_data['Cpt Type'].nunique()

In [633]:
# confirm the number of rows is same after merger
assert df_data.shape[0] == df_data_clean.shape[0]

In [634]:
# mirro allowance column as cost
df_data_clean['Cost'] = df_data_clean['Medicare Allowable']


In [635]:
# generate cost (If MCA is greater that zero use % of medicare allowance else use % of revenue)
# calculate cost value and replace
for ind in df_data_clean.index:
    if df_data_clean['Medicare Allowable'][ind] > 0:
        df_data_clean['Cost'][ind] = (df_data_clean['% of MCA'][ind] * df_data_clean['Medicare Allowable'][ind])/100
        df_data_clean['% of Revenue'][ind] = 0
    elif df_data_clean['Medicare Allowable'][ind] <= 0:
        df_data_clean['Cost'][ind] = (df_data_clean['% of Revenue'][ind] * df_data_clean['Revenue'][ind])/100
        df_data_clean['% of MCA'][ind] = 0

C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  if __name__ == '__main__':
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:5: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  """
C:\ProgramData\Anaconda3\lib\site-packages\ipykernel_launcher.py:

##### Test

In [636]:
df_data_clean[['Revenue','Medicare Allowable','% of MCA', '% of Revenue', 'Cost']]

Revenue  Medicare Allowable  % of MCA  % of Revenue         Cost
0      1823.640000            0.000000       0.0          90.0  1641.276000
1      1823.640000            0.000000       0.0          90.0  1641.276000
2      1823.640000            0.000000       0.0          90.0  1641.276000
3      1823.640000            0.000000       0.0          90.0  1641.276000
4      1823.640000            0.000000       0.0          90.0  1641.276000
...            ...                 ...       ...           ...          ...
10315    43.546866           31.104904      75.0           0.0    23.328678
10316    43.546866           31.104904      75.0           0.0    23.328678
10317   212.734334          151.953095      65.0           0.0    98.769512
10318   802.262136          573.044383      65.0           0.0   372.478849
10319   805.462000          573.044383      65.0           0.0   372.478849

[10320 rows x 5 columns]

In [637]:
df_data_clean.columns

Index(['Person Number', 'charge DOS', 'Contract', 'COB1', 'Location',
       'Physician', 'Patient Age', 'Cpt Type', 'Co-Payment',
       'Encounter Number', 'Patient Type', 'Patient Zip Code',
       'Patient Gender', 'CPT', 'Revenue', 'Code Desc', 'Medicare Allowable',
       'Key', 'COB 1', 'Fin Class', 'LOB', 'TPA', 'Primary Key', '% of MCA',
       '% of Revenue', 'Cost'],
      dtype='object')

In [638]:
### Drop redudant columns from clean data
df_data_clean = df_data_clean.drop(['COB 1', 'Key', ], axis = 1)

## Storing Data
- Save cleaned `df_data_clean` master dataset to a excel file named "twc_case_study_master.xlsx".

In [639]:
df_data_clean.to_excel('twc_case_study_master.xlsx', sheet_name='master Data', index=False)

<a id='eda'></a>
## Exploratory Data Analysis

In [640]:
# load dataset
df_master = pd.read_excel('twc_case_study_master.xlsx')

In [641]:
df_master.head(3)

Person Number charge DOS    Contract    COB1    Location    Physician  \
0          21913 2020-01-06  Contract 1  COB1 4  Location 4  Physician 4   
1          18623 2019-05-09  Contract 1  COB1 4  Location 4  Physician 4   
2          24780 2019-01-11  Contract 1  COB1 4  Location 4  Physician 4   

   Patient Age  Cpt Type  Co-Payment  Encounter Number  ...  Revenue  \
0           73  Lucentis           0                 5  ...  1823.64   
1           69  Lucentis           0              1599  ...  1823.64   
2           75  Lucentis           0              1614  ...  1823.64   

         Code Desc Medicare Allowable Fin Class       LOB  TPA  \
0  Lucentis Delive                0.0       FFS  Medicare  NaN   
1  Lucentis Delive                0.0       FFS  Medicare  NaN   
2  Lucentis Delive                0.0       FFS  Medicare  NaN   

        Primary Key % of MCA % of Revenue      Cost  
0  Contract 1COB1 4      0.0         90.0  1641.276  
1  Contract 1COB1 4      0.0         90.0  1641.276  
2  Contract 1COB1 4      0.0         90.0  1641.276  

[3 rows x 24 columns]

### Which Patient Type in Location 10 brought the Most revenue in 2019?

In [642]:
df_master.columns

Index(['Person Number', 'charge DOS', 'Contract', 'COB1', 'Location',
       'Physician', 'Patient Age', 'Cpt Type', 'Co-Payment',
       'Encounter Number', 'Patient Type', 'Patient Zip Code',
       'Patient Gender', 'CPT', 'Revenue', 'Code Desc', 'Medicare Allowable',
       'Fin Class', 'LOB', 'TPA', 'Primary Key', '% of MCA', '% of Revenue',
       'Cost'],
      dtype='object')

In [643]:
#extract year from date
df_master['year'] = pd.DatetimeIndex(df_master['charge DOS']).year

In [644]:
#extract month from date
df_master['month'] = pd.DatetimeIndex(df_master['charge DOS']).month
df_master['month'] = df_master['month'].apply(lambda x: calendar.month_abbr[x])

In [645]:
loc_10 = df_master[df_master['Location'] == 'Location 10']

In [646]:
loc_10_2019 = loc_10[loc_10['year'] == 2019]

In [647]:
loc_10_2019

Person Number charge DOS     Contract      COB1     Location  \
139             9363 2019-03-14  Contract 12   COB1 18  Location 10   
140             9363 2019-05-09  Contract 12   COB1 18  Location 10   
141             9363 2019-02-14  Contract 12   COB1 18  Location 10   
167             9363 2019-09-04  Contract 12   COB1 18  Location 10   
176             9777 2019-06-26  Contract 43   COB1 59  Location 10   
...              ...        ...          ...       ...          ...   
10132          22936 2019-12-09   Contract 5    COB1 9  Location 10   
10187          20696 2019-01-15  Contract 12   COB1 18  Location 10   
10206          20160 2019-08-27  Contract 66  COB1 124  Location 10   
10210          25436 2019-02-14   Contract 8   COB1 13  Location 10   
10218          20160 2019-04-23  Contract 66  COB1 124  Location 10   

          Physician  Patient Age     Cpt Type  Co-Payment  Encounter Number  \
139    Physician 18           74        Eylea           0              1553   
140    Physician 18           74        Eylea           0              1561   
141    Physician 18           74        Eylea           0              1567   
167    Physician 12           74        Eylea           0              4188   
176    Physician 12           72        Eylea           0              4946   
...             ...          ...          ...         ...               ...   
10132  Physician 19           59        Other           0             32137   
10187  Physician 22           60        Other           0               114   
10206  Physician 22           58  Office Code           0              9451   
10210  Physician 18           57  Office Code           0             13090   
10218  Physician 22           58  Office Code           0             25629   

       ... Medicare Allowable  Fin Class         LOB     TPA  \
139    ...           0.000000        FFS    Medicare   TPA 6   
140    ...           0.000000        FFS    Medicare   TPA 6   
141    ...           0.000000        FFS    Medicare   TPA 6   
167    ...           0.000000        FFS    Medicare   TPA 6   
176    ...           0.000000        CAP    Medicare     NaN   
...    ...                ...        ...         ...     ...   
10132  ...           0.000000   Self-Pay  Commercial     NaN   
10187  ...          20.830000        FFS    Medicare   TPA 6   
10206  ...          47.473393        CAP    Medicaid  TPA 16   
10210  ...          47.473393        CAP  Commercial   TPA 2   
10218  ...          47.473393        CAP    Medicaid  TPA 16   

               Primary Key % of MCA  % of Revenue         Cost  year month  
139     Contract 12COB1 18      0.0          90.0  2202.795000  2019   Mar  
140     Contract 12COB1 18      0.0          90.0  2202.795000  2019   May  
141     Contract 12COB1 18      0.0          90.0  2202.795000  2019   Feb  
167     Contract 12COB1 18      0.0          90.0  2202.795000  2019   Sep  
176     Contract 43COB1 59      0.0          90.0  2202.795000  2019   Jun  
...                    ...      ...           ...          ...   ...   ...  
10132     Contract 5COB1 9      0.0          90.0     0.000000  2019   Dec  
10187   Contract 12COB1 18     65.0           0.0    13.539500  2019   Jan  
10206  Contract 66COB1 124     75.0           0.0    35.605044  2019   Aug  
10210    Contract 8COB1 13     75.0           0.0    35.605044  2019   Feb  
10218  Contract 66COB1 124     75.0           0.0    35.605044  2019   Apr  

[318 rows x 26 columns]

In [648]:
loc_10_2019.groupby('Patient Type')['Revenue'].sum().sort_values(ascending = False)

Patient Type
Retinal Injections    58951.410
Office Visit          11806.606
Diagnostic Test         762.132
Office Procedure        729.652
Post Op                 243.558
Visual Field            157.416
Post Op 1 Day             0.000
0                         0.000
Name: Revenue, dtype: float64

>> Which Patient Type in Location 10 brought the Most revenue in 2019?
Ans : Retinal Injections had the highest revenue in 2019 from location 10

### Which location saw the biggest increase in total patient volume %? what %? (1/1/2020-3/16/2020 vs 3/17/2020-6/30/2020).

In [649]:
# records above '2020-01-01'
ll = df_master[df_master['charge DOS'] >= '2020-01-01']

In [650]:
# records between '2020-01-01' and '2020-03-16'
lH = ll[ll['charge DOS'] <= '2020-03-16']

In [651]:
# records above '2020-03-17'
Hl = df_master[df_master['charge DOS'] >= '2020-03-17']

In [652]:
# records above '2020-06-30'
HH = Hl[Hl['charge DOS'] >= '2020-06-30']

In [653]:
# Group by patient
below = lH.groupby('Location')['Person Number'].count().sort_values(ascending = False)

In [654]:
#convert to dataframe
df_bl = pd.DataFrame(below)
df_bl

Person Number
Location                  
Location 4             278
Location 6              86
Location 35             64
Location 18             63
Location 31             48
Location 10             36
Location 14             32
Location 23             27
Location 8              13
Location 29              6
Location 44              2
Location 45              1

In [655]:
# group upper end by patient
above = HH.groupby('Location')['Person Number'].count().sort_values(ascending = False)

In [656]:
#convert to dataframe
df_hl = pd.DataFrame(above)
df_hl

Person Number
Location                  
Location 4             795
Location 18            221
Location 6             207
Location 10            178
Location 35            154
Location 31            145
Location 23            118
Location 8              88
Location 14             29
Location 41              9
Location 29              8
Location 44              5

In [657]:
#rename for merger
df_hl = df_hl.rename(columns = {'Person Number' : 'Person Number_2'})

In [658]:
# join both upper and lower patient count 
Dff = df_bl.join(df_hl)

In [659]:
Dff

Person Number  Person Number_2
Location                                   
Location 4             278            795.0
Location 6              86            207.0
Location 35             64            154.0
Location 18             63            221.0
Location 31             48            145.0
Location 10             36            178.0
Location 14             32             29.0
Location 23             27            118.0
Location 8              13             88.0
Location 29              6              8.0
Location 44              2              5.0
Location 45              1              NaN

In [660]:
Dff["% increase"] = ((Dff["Person Number_2"] - Dff["Person Number"])/Dff["Person Number"])*100

In [661]:
Dff.sort_values(by = ['% increase'],ascending=False)

Person Number  Person Number_2  % increase
Location                                               
Location 8              13             88.0  576.923077
Location 10             36            178.0  394.444444
Location 23             27            118.0  337.037037
Location 18             63            221.0  250.793651
Location 31             48            145.0  202.083333
Location 4             278            795.0  185.971223
Location 44              2              5.0  150.000000
Location 6              86            207.0  140.697674
Location 35             64            154.0  140.625000
Location 29              6              8.0   33.333333
Location 14             32             29.0   -9.375000
Location 45              1              NaN         NaN

>> Which location saw the biggest increase in total patient volume %? Location 8
what %? 576.923077  (1/1/2020-3/16/2020 vs 3/17/2020-6/30/2020)

### Do females or males bring in more revenue/patient?

In [662]:
df_master.groupby('Patient Gender')['Revenue'].sum().sort_values(ascending = False)

Patient Gender
M    1.618320e+06
F    1.324511e+06
0    1.982981e+03
Name: Revenue, dtype: float64

>> Male Gender generate more revenue compared to the female gender

### What is the cost/patient of our most costly patient type? Which patient type?

In [663]:
df_cost = df_master.groupby('Patient Type')['Cost'].sum().sort_values(ascending = False)
df_cost

Patient Type
Retinal Injections      1.546593e+06
Surgery                 2.331057e+05
Office Visit            2.053919e+05
Retinal Procedures      3.150655e+04
Retinal Lasers          2.616829e+04
Anesthesia              1.590946e+04
Office Procedure        1.166232e+04
Post Op                 1.021910e+04
Yag Laser ASC           5.485130e+03
O C T Test              5.366909e+03
Diagnostic Test         4.296648e+03
SLT Laser ASC           3.837319e+03
0                       3.756277e+03
Visual Field            3.436524e+03
SLT Lasers              2.895162e+03
Medication              2.267554e+03
Yag Laser               2.240119e+03
Emergency Call Visit    2.157194e+03
Yag Laser Office        1.924682e+03
New Pt                  1.499375e+03
SLT Laser Office        4.796649e+02
Post Op 1 Day           1.676971e+02
Shared Post Op          1.262358e+02
Medical Records Fee     9.772199e+01
Telemedicine            4.414083e+01
Walk In                 2.800504e+01
Pre Op                  0

In [664]:
df_pat_count = df_master.groupby('Patient Type')['Person Number'].count().sort_values(ascending = False)
df_pat_count

Patient Type
Retinal Injections      5382
Office Visit            2605
Post Op                  604
Surgery                  364
Anesthesia               246
Post Op 1 Day            220
Retinal Procedures       128
Diagnostic Test          121
Retinal Lasers           108
Visual Field              92
O C T Test                75
Office Procedure          60
Medication                59
Emergency Call Visit      49
SLT Laser ASC             49
SLT Lasers                38
0                         28
Yag Laser ASC             28
New Pt                    25
Yag Laser                 11
SLT Laser Office           6
Medical Records Fee        6
Yag Laser Office           5
Walk In                    4
Shared Post Op             3
Telemedicine               3
Pre Op                     1
Name: Person Number, dtype: int64

In [665]:
df_cost = pd.DataFrame(df_cost)
df_pat_count = pd.DataFrame(df_pat_count)
df_merger = df_cost.join(df_pat_count)

In [666]:
df_merger['cost/patient'] = df_merger['Cost'] / df_merger['Person Number']

In [667]:
df_merger.sort_values(by = 'cost/patient', ascending = False)

Cost  Person Number  cost/patient
Patient Type                                                   
Surgery               2.331057e+05            364    640.400335
Yag Laser Office      1.924682e+03              5    384.936462
Retinal Injections    1.546593e+06           5382    287.363950
Retinal Procedures    3.150655e+04            128    246.144913
Retinal Lasers        2.616829e+04            108    242.299021
Yag Laser             2.240119e+03             11    203.647150
Yag Laser ASC         5.485130e+03             28    195.897498
Office Procedure      1.166232e+04             60    194.371950
0                     3.756277e+03             28    134.152749
SLT Laser Office      4.796649e+02              6     79.944150
Office Visit          2.053919e+05           2605     78.845252
SLT Laser ASC         3.837319e+03             49     78.312636
SLT Lasers            2.895162e+03             38     76.188481
O C T Test            5.366909e+03             75     71.558791
Anesthesia            1.590946e+04            246     64.672605
New Pt                1.499375e+03             25     59.974985
Emergency Call Visit  2.157194e+03             49     44.024373
Shared Post Op        1.262358e+02              3     42.078598
Medication            2.267554e+03             59     38.433120
Visual Field          3.436524e+03             92     37.353517
Diagnostic Test       4.296648e+03            121     35.509484
Post Op               1.021910e+04            604     16.919034
Medical Records Fee   9.772199e+01              6     16.286999
Telemedicine          4.414083e+01              3     14.713610
Walk In               2.800504e+01              4      7.001261
Post Op 1 Day         1.676971e+02            220      0.762259
Pre Op                0.000000e+00              1      0.000000

>> What is the cost/patient of our most costly patient type? 640.400335
Which patient type? Surgery

### Which Physician was the most profitable on a per patient basis? What profit?

In [668]:
df_master.columns

Index(['Person Number', 'charge DOS', 'Contract', 'COB1', 'Location',
       'Physician', 'Patient Age', 'Cpt Type', 'Co-Payment',
       'Encounter Number', 'Patient Type', 'Patient Zip Code',
       'Patient Gender', 'CPT', 'Revenue', 'Code Desc', 'Medicare Allowable',
       'Fin Class', 'LOB', 'TPA', 'Primary Key', '% of MCA', '% of Revenue',
       'Cost', 'year', 'month'],
      dtype='object')

In [669]:
# calculate the profit
df_master["Profit"] = df_master['Revenue'] - df_master['Cost']

In [670]:
df_phy_rev = df_master.groupby('Physician')['Profit'].sum().sort_values(ascending = False)
df_phy_rev

Physician
Physician 4      262951.429958
Physician 14     217674.480227
Physician 18     101738.966603
Physician 12      41171.842644
Physician 48      34478.779904
Physician 29      32834.800632
Physician 16      28167.755200
Physician 22      22173.859991
Physician 7       19883.919944
Physician 19      15441.376723
Physician 13      10423.551266
Physician 41       9811.313764
Physician 27       5847.135408
Physician 15       4735.600669
Physician 10       1812.571797
Physician 68       1735.760696
Physician 36       1687.525643
Physician 58       1602.032372
Physician 43       1082.484359
Physician 82       1046.348344
Physician 67        902.381067
Physician 49        640.510111
Physician 60        556.602936
Physician 53        447.294869
Physician 90        388.131184
Physician 70        372.129881
Physician 71        330.697255
Physician 78        231.454143
Physician 25        225.546931
Physician 65        192.004929
Physician 94        150.928192
Physician 66        149.03553

In [671]:
df_phy_pat_count = df_master.groupby('Physician')['Person Number'].count().sort_values(ascending = False)
df_phy_pat_count

Physician
Physician 4      3497
Physician 14     2121
Physician 18     1521
Physician 12      540
Physician 29      536
Physician 16      420
Physician 22      323
Physician 7       321
Physician 48      245
Physician 13      157
Physician 19      120
Physician 41       72
Physician 27       53
Physician 10       51
Physician 82       47
Physician 15       38
Physician 36       35
Physician 58       31
Physician 68       30
Physician 60       18
Physician 67       16
Physician 71       14
Physician 53       12
Physician 90       11
Physician 66       11
Physician 49       11
Physician 43       10
Physician 70        8
Physician 25        7
Physician 78        7
Physician 94        7
Physician 31        6
Physician 65        5
Physician 84        5
Physician 73        3
Physician 79        3
Physician 102       3
Physician 87        2
Physician 99        2
Physician 105       1
Name: Person Number, dtype: int64

In [672]:
df_phy_rev = pd.DataFrame(df_phy_rev)
df_phy_pat_count = pd.DataFrame(df_phy_pat_count)
df_phy_merger = df_phy_rev.join(df_phy_pat_count)
df_phy_merger

Profit  Person Number
Physician                                  
Physician 4    262951.429958           3497
Physician 14   217674.480227           2121
Physician 18   101738.966603           1521
Physician 12    41171.842644            540
Physician 48    34478.779904            245
Physician 29    32834.800632            536
Physician 16    28167.755200            420
Physician 22    22173.859991            323
Physician 7     19883.919944            321
Physician 19    15441.376723            120
Physician 13    10423.551266            157
Physician 41     9811.313764             72
Physician 27     5847.135408             53
Physician 15     4735.600669             38
Physician 10     1812.571797             51
Physician 68     1735.760696             30
Physician 36     1687.525643             35
Physician 58     1602.032372             31
Physician 43     1082.484359             10
Physician 82     1046.348344             47
Physician 67      902.381067             16
Physician 49      640.510111             11
Physician 60      556.602936             18
Physician 53      447.294869             12
Physician 90      388.131184             11
Physician 70      372.129881              8
Physician 71      330.697255             14
Physician 78      231.454143              7
Physician 25      225.546931              7
Physician 65      192.004929              5
Physician 94      150.928192              7
Physician 66      149.035532             11
Physician 79      142.305429              3
Physician 84      123.002592              5
Physician 87      119.455007              2
Physician 31      117.676356              6
Physician 73      107.412549              3
Physician 102      70.880285              3
Physician 105      53.338627              1
Physician 99       47.863611              2

In [673]:
df_phy_merger['Profit/patient'] = df_phy_merger['Profit'] / df_phy_merger['Person Number']

In [674]:
df_phy_merger.sort_values(by = 'Profit/patient', ascending = False)

Profit  Person Number  Profit/patient
Physician                                                  
Physician 48    34478.779904            245      140.729714
Physician 41     9811.313764             72      136.268247
Physician 19    15441.376723            120      128.678139
Physician 15     4735.600669             38      124.621070
Physician 27     5847.135408             53      110.323310
Physician 43     1082.484359             10      108.248436
Physician 14   217674.480227           2121      102.628232
Physician 12    41171.842644            540       76.244153
Physician 4    262951.429958           3497       75.193432
Physician 22    22173.859991            323       68.649721
Physician 16    28167.755200            420       67.066084
Physician 18   101738.966603           1521       66.889524
Physician 13    10423.551266            157       66.392046
Physician 7     19883.919944            321       61.943676
Physician 29    32834.800632            536       61.258956
Physician 87      119.455007              2       59.727504
Physician 49      640.510111             11       58.228192
Physician 68     1735.760696             30       57.858690
Physician 67      902.381067             16       56.398817
Physician 105      53.338627              1       53.338627
Physician 58     1602.032372             31       51.678464
Physician 36     1687.525643             35       48.215018
Physician 79      142.305429              3       47.435143
Physician 70      372.129881              8       46.516235
Physician 65      192.004929              5       38.400986
Physician 53      447.294869             12       37.274572
Physician 73      107.412549              3       35.804183
Physician 10     1812.571797             51       35.540623
Physician 90      388.131184             11       35.284653
Physician 78      231.454143              7       33.064878
Physician 25      225.546931              7       32.220990
Physician 60      556.602936             18       30.922385
Physician 84      123.002592              5       24.600518
Physician 99       47.863611              2       23.931806
Physician 102      70.880285              3       23.626762
Physician 71      330.697255             14       23.621232
Physician 82     1046.348344             47       22.262731
Physician 94      150.928192              7       21.561170
Physician 31      117.676356              6       19.612726
Physician 66      149.035532             11       13.548685

>> Which Physician was the most profitable on a per patient basis? Physician 48
What profit? 140.729714

### Which doctor saw the highest volume of unique Medicare Patients.

In [675]:
df_master.groupby('Physician')['Person Number'].nunique().sort_values(ascending = False)

Physician
Physician 48     71
Physician 14     62
Physician 4      55
Physician 18     35
Physician 16     23
Physician 19     22
Physician 82     19
Physician 13     18
Physician 22     18
Physician 27     18
Physician 29     17
Physician 41     16
Physician 10     16
Physician 15     14
Physician 12     13
Physician 7      12
Physician 58     11
Physician 36     11
Physician 60     10
Physician 71      9
Physician 53      7
Physician 66      6
Physician 49      6
Physician 68      6
Physician 43      5
Physician 25      4
Physician 67      4
Physician 31      4
Physician 78      4
Physician 84      3
Physician 70      3
Physician 79      2
Physician 73      2
Physician 87      2
Physician 90      2
Physician 65      2
Physician 102     2
Physician 94      1
Physician 105     1
Physician 99      1
Name: Person Number, dtype: int64

>> Which doctor saw the highest volume of unique Medicare Patients. Physician 48

### What is the average age of a female Glaucoma & Glaucoma Laser patient in 2019?

In [676]:
df_master.columns

Index(['Person Number', 'charge DOS', 'Contract', 'COB1', 'Location',
       'Physician', 'Patient Age', 'Cpt Type', 'Co-Payment',
       'Encounter Number', 'Patient Type', 'Patient Zip Code',
       'Patient Gender', 'CPT', 'Revenue', 'Code Desc', 'Medicare Allowable',
       'Fin Class', 'LOB', 'TPA', 'Primary Key', '% of MCA', '% of Revenue',
       'Cost', 'year', 'month', 'Profit'],
      dtype='object')

In [677]:
df_master['Cpt Type'].unique()

array(['Lucentis', 'Eylea', 'Other', 'Avastin', 'OCT', 'Office Code',
       'Post-op', 'Retinal Laser', 'SLT Laser', 'Visual Field',
       'Plastic Surgery', 'Other Injectables', 'Ozurdex',
       'Cataract Surgery', 'Glaucoma', 'Glaucoma Laser',
       'Non-Cataract Surgery', 'Yag & PI Laser', 'Retinal Surgery',
       'Anesthesia', 'Ophthalmoscopy', 'Beovu', 0, 'Kenalog'],
      dtype=object)

In [678]:
df_gla_las = df_master[((df_master['Cpt Type'] == 'Glaucoma') | (df_master['Cpt Type'] == 'Glaucoma Laser')) & (df_master['Patient Gender'] == 'F')]
df_gla_las_2019 = df_gla_las [df_gla_las ['year'] == 2019]

In [679]:
df_gla_las_2019.describe()

Person Number  Patient Age  Co-Payment  Encounter Number  \
count        8.00000     8.000000    8.000000          8.000000   
mean     13718.00000    78.625000    1.875000       3470.625000   
std      10322.20937    14.657397    2.587746       2059.295229   
min       1220.00000    57.000000    0.000000        998.000000   
25%       7273.00000    72.000000    0.000000       1010.750000   
50%       8191.00000    81.000000    0.000000       4694.500000   
75%      25436.00000    92.000000    5.000000       5102.000000   
max      26724.00000    92.000000    5.000000       5137.000000   

       Patient Zip Code      Revenue  Medicare Allowable  % of MCA  \
count          8.000000     8.000000            8.000000       8.0   
mean       90048.250000  1446.053000         1034.062366      75.0   
std            5.775564   527.643161          373.349277       0.0   
min        90045.000000   582.274000          429.910000      75.0   
25%        90045.000000  1405.680500         1000.561697      75.0   
50%        90045.000000  1742.132000         1244.380000      75.0   
75%        90048.500000  1742.132000         1244.380000      75.0   
max        90059.000000  1742.132000         1244.380000      75.0   

       % of Revenue        Cost    year      Profit  
count           8.0    8.000000     8.0    8.000000  
mean            0.0  775.546775  2019.0  670.506225  
std             0.0  280.011958     0.0  247.667292  
min             0.0  322.432500  2019.0  259.841500  
25%             0.0  750.421273  2019.0  655.259227  
50%             0.0  933.285000  2019.0  808.847000  
75%             0.0  933.285000  2019.0  808.847000  
max             0.0  933.285000  2019.0  808.847000

>> What is the average age of a female Glaucoma & Glaucoma Laser patient in 2019? 78.625000

## Which patient type drives the most revenue/encounter and profit/encounter? What was Revenue by Encounter/ Profit/Encounter?

In [680]:
df_master.columns

Index(['Person Number', 'charge DOS', 'Contract', 'COB1', 'Location',
       'Physician', 'Patient Age', 'Cpt Type', 'Co-Payment',
       'Encounter Number', 'Patient Type', 'Patient Zip Code',
       'Patient Gender', 'CPT', 'Revenue', 'Code Desc', 'Medicare Allowable',
       'Fin Class', 'LOB', 'TPA', 'Primary Key', '% of MCA', '% of Revenue',
       'Cost', 'year', 'month', 'Profit'],
      dtype='object')

In [681]:
df_master['Revenue/Encounter_Number'] = df_master['Revenue'] / df_master['Encounter Number']

In [682]:
df_master['Profit/Encounter_Number'] = df_master['Profit'] / df_master['Encounter Number']

In [683]:
df_revenue_encounter = df_master.groupby('Patient Type')['Revenue/Encounter_Number'].sum().sort_values(ascending = False)
df_revenue_encounter = pd.DataFrame(df_revenue_encounter)
df_revenue_encounter

Revenue/Encounter_Number
Patient Type                                  
Retinal Injections                  980.062296
Surgery                             106.028061
Office Visit                         61.147962
Retinal Lasers                       19.373765
Retinal Procedures                   10.889163
Anesthesia                            8.583825
SLT Lasers                            3.753944
Office Procedure                      3.143667
Yag Laser ASC                         3.095220
SLT Laser ASC                         2.141021
New Pt                                1.822599
Post Op                               1.801638
Yag Laser                             1.775182
0                                     1.571674
Visual Field                          1.412311
Diagnostic Test                       1.365794
O C T Test                            0.550116
Emergency Call Visit                  0.470964
Medication                            0.203394
Yag Laser Office                      0.113895
SLT Laser Office                      0.055220
Post Op 1 Day                         0.050593
Shared Post Op                        0.037484
Medical Records Fee                   0.010307
Telemedicine                          0.002917
Walk In                               0.001811
Pre Op                                0.000000

In [684]:
df_profit_encounter = df_master.groupby('Patient Type')['Profit/Encounter_Number'].sum().sort_values(ascending = False)
df_profit_encounter = pd.DataFrame(df_profit_encounter)
df_profit_encounter

Profit/Encounter_Number
Patient Type                                 
Retinal Injections                 164.914782
Surgery                             49.705737
Office Visit                        20.104764
Retinal Lasers                       9.239030
Anesthesia                           5.906538
Retinal Procedures                   2.169168
SLT Lasers                           2.030322
Yag Laser ASC                        1.865341
Office Procedure                     1.845579
SLT Laser ASC                        1.159650
Yag Laser                            1.073877
Post Op                              0.919295
New Pt                               0.877403
Diagnostic Test                      0.692905
Visual Field                         0.685517
0                                    0.485546
O C T Test                           0.243209
Emergency Call Visit                 0.224469
Medication                           0.081679
SLT Laser Office                     0.034064
Yag Laser Office                     0.029530
Post Op 1 Day                        0.024456
Shared Post Op                       0.017717
Medical Records Fee                  0.005621
Telemedicine                         0.001468
Walk In                              0.000972
Pre Op                               0.000000

In [685]:
df_merger_encounter = df_revenue_encounter.join(df_profit_encounter)
df_merger_encounter

Revenue/Encounter_Number  Profit/Encounter_Number
Patient Type                                                           
Retinal Injections                  980.062296               164.914782
Surgery                             106.028061                49.705737
Office Visit                         61.147962                20.104764
Retinal Lasers                       19.373765                 9.239030
Retinal Procedures                   10.889163                 2.169168
Anesthesia                            8.583825                 5.906538
SLT Lasers                            3.753944                 2.030322
Office Procedure                      3.143667                 1.845579
Yag Laser ASC                         3.095220                 1.865341
SLT Laser ASC                         2.141021                 1.159650
New Pt                                1.822599                 0.877403
Post Op                               1.801638                 0.919295
Yag Laser                             1.775182                 1.073877
0                                     1.571674                 0.485546
Visual Field                          1.412311                 0.685517
Diagnostic Test                       1.365794                 0.692905
O C T Test                            0.550116                 0.243209
Emergency Call Visit                  0.470964                 0.224469
Medication                            0.203394                 0.081679
Yag Laser Office                      0.113895                 0.029530
SLT Laser Office                      0.055220                 0.034064
Post Op 1 Day                         0.050593                 0.024456
Shared Post Op                        0.037484                 0.017717
Medical Records Fee                   0.010307                 0.005621
Telemedicine                          0.002917                 0.001468
Walk In                               0.001811                 0.000972
Pre Op                                0.000000                 0.000000

>> Which patient type drives the most revenue/encounter and profit/encounter? Retinal Injections	
What was Revenue by Encounter/ Profit/Encounter? 980.062296	/ 164.914782

### Which contract type is most profitable per patient/what profit?

In [686]:
df_master.columns

Index(['Person Number', 'charge DOS', 'Contract', 'COB1', 'Location',
       'Physician', 'Patient Age', 'Cpt Type', 'Co-Payment',
       'Encounter Number', 'Patient Type', 'Patient Zip Code',
       'Patient Gender', 'CPT', 'Revenue', 'Code Desc', 'Medicare Allowable',
       'Fin Class', 'LOB', 'TPA', 'Primary Key', '% of MCA', '% of Revenue',
       'Cost', 'year', 'month', 'Profit', 'Revenue/Encounter_Number',
       'Profit/Encounter_Number'],
      dtype='object')

In [687]:
df_contract_profit = df_master.groupby('LOB')['Profit'].sum().sort_values(ascending = False)
df_contract_profit = pd.DataFrame(df_contract_profit)
df_contract_profit

Profit
LOB                      
Medicare    674774.886249
Commercial  105949.914257
Medicaid     40945.387124

In [688]:
df_contract_patient = df_master.groupby('LOB')['Person Number'].count().sort_values(ascending = False)
df_contract_patient = pd.DataFrame(df_contract_patient)
df_contract_patient

Person Number
LOB                      
Medicare             8301
Commercial           1393
Medicaid              626

In [689]:
df_merger_contract = df_contract_profit.join(df_contract_patient)
df_merger_contract 

Profit  Person Number
LOB                                     
Medicare    674774.886249           8301
Commercial  105949.914257           1393
Medicaid     40945.387124            626

In [690]:
df_merger_contract['Profit/Person_Number'] = df_merger_contract['Profit'] / df_merger_contract['Person Number']

In [691]:
df_merger_contract.sort_values(by = 'Profit/Person_Number', ascending = False)

Profit  Person Number  Profit/Person_Number
LOB                                                           
Medicare    674774.886249           8301             81.288385
Commercial  105949.914257           1393             76.058804
Medicaid     40945.387124            626             65.407967

### What is the total profit on FFS patients?

In [692]:
df_master.groupby('Fin Class')['Profit'].sum().sort_values(ascending = False)

Fin Class
CAP         486450.471553
FFS         334268.208324
Self-Pay       951.507753
Name: Profit, dtype: float64

<a id='conclusions'></a>
## Conclusions

